OrbitService/ProducerSideServiceImpl.cpp

```diff
@@ -0,0 +1,324 @@
+  CHECK(capture_event_buffer != nullptr);
+  LOG("About to send StartCaptureCommand to CaptureEventProducers (if any)");
+  {
+    absl::WriterMutexLock lock{&capture_event_buffer_mutex_};
+    capture_event_buffer_ = capture_event_buffer;
+  }
+  {
+    absl::MutexLock lock{&service_state_mutex_};
+    service_state_.capture_status = CaptureStatus::kCaptureStarted;
+  }
+}
+
+void ProducerSideServiceImpl::OnCaptureStopRequested() {
+  LOG("About to send StopCaptureCommand to CaptureEventProducers (if any)");
+  {
+    absl::MutexLock lock{&service_state_mutex_};
+    service_state_.capture_status = CaptureStatus::kCaptureStopping;
+
+    // Wait (for a limited amount of time) for all producers to send AllEventsSent or to disconnect.
+    service_state_mutex_.AwaitWithTimeout(absl::Condition(
+                                              +[](ServiceState* service_state) {
+                                                return service_state->producers_remaining == 0 ||
+                                                       service_state->exit_requested;
+                                              },
+                                              &service_state_),
+                                          max_wait_for_all_events_sent_);
+    CHECK(service_state_.producers_remaining >= 0);
+    if (service_state_.producers_remaining == 0) {
+      LOG("All CaptureEventProducers have finished sending their CaptureEvents");
+    } else {
+      ERROR(
+          "Stopped receiving CaptureEvents from CaptureEventProducers "
+          "even if not all have sent all their CaptureEvents");
+    }
+    service_state_.capture_status = CaptureStatus::kCaptureFinished;
+    service_state_.producers_remaining = 0;
+  }
+
+  {
+    absl::WriterMutexLock lock{&capture_event_buffer_mutex_};
+    capture_event_buffer_ = nullptr;
+  }
+}
+
+void ProducerSideServiceImpl::OnExitRequest() {
+  {
+    absl::MutexLock lock{&service_state_mutex_};
+    service_state_.exit_requested = true;
+  }
+
+  LOG("Attempting to disconnect from CaptureEventProducers as exit was requested");
+  {
+    absl::MutexLock lock{&server_contexts_mutex_};
+    for (grpc::ServerContext* context : server_contexts_) {
+      // This should cause blocking Reads on ServerReaderWriter to fail immediately.
+      context->TryCancel();
+    }
+  }
+
+  {
+    absl::WriterMutexLock lock{&capture_event_buffer_mutex_};
+    capture_event_buffer_ = nullptr;
+  }
+}
+
+grpc::Status ProducerSideServiceImpl::ReceiveCommandsAndSendEvents(
+    ::grpc::ServerContext* context,
+    ::grpc::ServerReaderWriter< ::orbit_grpc_protos::ReceiveCommandsAndSendEventsResponse,
+                                ::orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest>* stream) {
+  LOG("A CaptureEventProducer has connected calling ReceiveCommandsAndSendEvents");
+
+  {
+    absl::MutexLock lock{&server_contexts_mutex_};
+    server_contexts_.emplace(context);
+  }
```

Comment:
SetContext?

Reply:
As above, I don't want to add more methods that take mutexes, even more here are there's no code duplication.